### PR TITLE
Step12

### PR DESCRIPTION
**Pull request by mohammad-suliman**
This PR implements step 12 of the sensory interface project. Note the following:
* Possibly, this code is not final, it may change according to @oferb's feedback
* The PR also adjusts the "updateURL" function to work also on local host. I have done this to make life easier for me when testing the code :)

### CODE REVIEW

File: public/view/view.js [hunk 1]
```diff
@@ -121,6 +121,18 @@ function navigateGrid(event) {
 }
 
 function get2DCoordinates(currentCell) {
+    /*
+    * This function maps the row and col attributes for each cell in the grid to 2D coordinates (x,y)
```
Comment (oferb) [inline public/view/view.js]: General guideline - 
Documentation for what a method does should be outside it.
Documentation for how it does it should be inside it, preferably split to explain the line or lines just after it.

File: public/view/view.js [hunk 2]
```diff
@@ -121,6 +121,18 @@ function navigateGrid(event) {
...
+    * Row index is mapped to y coordinate, and col index is mapped to x coordinate
```
Comment (oferb) [inline public/view/view.js]: Try to stay within line lengths of 100 characters.

File: public/view/view.js [hunk 3]
```diff
@@ -11,22 +11,23 @@ let timeOut = null;
...
 
 function createGrid() {
     let grid = document.createElement("div");
     grid.setAttribute("role", "grid");
+    grid.setAttribute("id", "grid");
     grid.setAttribute("aria-readonly", "true");
     grid.style.width = "100%";
     grid.style.height = "70%";
-    grid.className = "table";
+    grid.className = "  table";
```
Comment (oferb) [inline public/view/view.js]: Did you add the spaces here by mistake?

File: public/view/view.js [hunk 4]
```diff
@@ -111,44 +113,105 @@ function navigateGrid(event) {
...
+    let maxCoordinates = get2DCoordinates(cellWithMaxCoordinates);
+    let maxDistance = 0;
+    maxDistance += Math.pow(maxCoordinates.x, 2);
+    maxDistance += Math.pow(maxCoordinates.y, 2);
+    maxDistance = Math.sqrt(maxDistance);
+    return maxDistance;
+}
+
+function createAndSetPanner(currentCell) {
+    let panner = audioContext.createPanner();
```
[the remark this answers is not in the document]
Comment (mohammad-suliman) [reply public/view/view.js]: no, but from what I read, it's only for left and right, no 3D capability.

File: public/view/view.js [hunk 5]
```diff
@@ -111,44 +113,105 @@ function navigateGrid(event) {
...
+}
+
+function playSound(event) {
+    if (audioContext.state == "suspended") {
+        audioContext.resume();
+    }
+    let currentCell = currentCellUnderTouchPoint;
+    createAndSetOscillator(currentCell);
+    let panner = createAndSetPanner(currentCell);
+    oscillator.connect(panner);
```
[the remark this answers is not in the document]
Comment (mohammad-suliman) [reply public/view/view.js]: Yes in fact, please have a look here:
https://blog.szynalski.com/2014/04/web-audio-api/

File: public/view/view.js (hunk 3 above)
Comment (mohammad-suliman) [reply public/view/view.js]: Correct :)

File: public/view/view.js [hunk 6]
```diff
@@ -111,44 +120,118 @@ function navigateGrid(event) {
...
+* to 2D coordinates (x,y)
+* Row index is mapped to y coordinate, and col index is mapped to x coordinate
+* The row and col attributes is zero based indexes. 
+* For example, the values of row and col attributes for a cell found in 
+* the second row, and in the third column is:
+* Row is 1, and col is 2
+*/
+function get2DCoordinates(currentCell) {
+    /**
+    * This function does the mapping from indexes to 2D coordinates as follows:
```
Comment (oferb) [inline public/view/view.js]: This comment adds new information, but also repeats the information in the code.
Try to shorten it by only adding new information not already in the code.

Also, see if you can put this comment closer to the relevant lines of code, possibly splitting it into pieces if needed.

File: public/view/view.js [hunk 7]
```diff
@@ -111,44 +120,118 @@ function navigateGrid(event) {
...
-function startSoundPlayback(event) {
-    currentCellUnderTouchPoint = event.currentTarget;
-    event.preventDefault();
-    stopSoundPlayback(event);
-    let selectedValue = event.currentTarget.firstChild.data;
-    playSound(selectedValue);
+/**
+* This function calculates the greatest distance possible for a cell in the grid, 
+* when it's coordinates is mapped to 2D coordinates
+* This cell is naturaly  found in the upper right corner of the grid
```
Comment (oferb) [inline public/view/view.js]: Nit: double space in `naturaly  found`.

File: public/view/view.js [hunk 8]
```diff
@@ -111,44 +120,118 @@ function navigateGrid(event) {
...
+*/
+function getMaxDistancePossible() {
+    /** 
+    * The function calculates distance of a given cell from the origin (0,0), 
+    * according to the wellknown distance formula
```
Comment (oferb) [inline public/view/view.js]: You can remove `according to the wellknown distance formula` and just add `the Euclidean`: `The function calculates the Euclidean distance`

File: public/view/view.js [hunk 9]
```diff
@@ -4,34 +4,39 @@ let audioContext = new AudioContext();
 let oscillator = null;
 // This variable is not used currently.
 let source = null;
+// This variable stores the current cell under touch point in case touch is available.
+// In case touch is not available, it stores the current focused cell.
 let currentCellUnderTouchPoint = null;
```
[the remark this answers is not in the document]
Comment (mohammad-suliman) [reply public/view/view.js]: What about currentCellUnderTouchOrFocus? very long?

[the remark this answers is not in the document]
Comment (oferb) [reply public/view/view.js]: Yah, very long. How about `selectedCell`?

[the remark this answers is not in the document]
Comment (mohammad-suliman) [reply public/view/view.js]: Sounds OK to me.

File: public/view/view.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (oferb) [reply public/view/view.js]: Interesting! Can you please document this?

[the remark this answers is not in the document]
Comment (mohammad-suliman) [reply public/view/view.js]: OK :)

File: public/view/view.js [hunk 10]
```diff
@@ -111,44 +120,115 @@ function navigateGrid(event) {
...
+* and the value is the calculated x coordinate,
+* and the second element's key is y, and the value is the calculated y coordinate
+*/
+function get2DCoordinates(currentCell) {
+    let grid = document.getElementById("grid");
+    let columnCount = grid.firstChild.childNodes.length;
+    let columnNumber = currentCell.getAttribute("col");
+    let xCoordinate = columnNumber - Math.floor(columnCount / 2);
+    // To keep the value of the x coordinate  symmetric with respect to  y-axis, 
+    // we increment the value of x coordinate by 0.5 if it's positive,and column count is even
```
Comment (oferb) [inline public/view/view.js]: > if it's positive
Also if it's non-positive.

File: public/view/view.js [hunk 11]
```diff
@@ -111,44 +120,115 @@ function navigateGrid(event) {
...
+* The return value's type  of the function is object, where the key of the first element is x, 
+* and the value is the calculated x coordinate,
+* and the second element's key is y, and the value is the calculated y coordinate
+*/
+function get2DCoordinates(currentCell) {
+    let grid = document.getElementById("grid");
+    let columnCount = grid.firstChild.childNodes.length;
+    let columnNumber = currentCell.getAttribute("col");
+    let xCoordinate = columnNumber - Math.floor(columnCount / 2);
+    // To keep the value of the x coordinate  symmetric with respect to  y-axis,
```
Comment (oferb) [inline public/view/view.js]: Some double-spaces on this line:
`// To keep the value of the x coordinate  symmetric with respect to  y-axis, `

File: public/view/view.js [hunk 12]
```diff
@@ -111,44 +120,115 @@ function navigateGrid(event) {
...
+    let xCoordinate = columnNumber - Math.floor(columnCount / 2);
+    // To keep the value of the x coordinate  symmetric with respect to  y-axis, 
+    // we increment the value of x coordinate by 0.5 if it's positive,and column count is even
+    if (columnCount % 2 == 0) {
+        xCoordinate += 0.5;
+    }
+    let rowCount = grid.childNodes.length;
+    let rowNumber = currentCell.getAttribute("row");
+    let yCoordinate = rowNumber - Math.floor(rowCount / 2);
+    // Keep y coordinate sytmetric with respect to x-axis by increnenting the value by 0.5
```
Comment (oferb) [inline public/view/view.js]: No need to repeat the whole comment. I would shorten to just:
`// Align y similar to x:`

File: public/view/view.js (hunk 10 above)
Comment (oferb) [inline public/view/view.js]: "we increment the value of x coordinate by 0.5 if it's positive,and column count is even"
This sentence repeats the code it describes, just in words:
```
if (columnCount % 2 == 0) {
  xCoordinate += 0.5;
}
```
So it's not needed.
Comments should add new information to the code. Think of it like a delta in information beyond what the code shows.
In some cases, where the code is very complex, the comments can describe the code, but these cases are rare (and should generally be avoided).

File: public/view/view.js (hunk 11 above)
Comment (oferb) [inline public/view/view.js]: Line comments should usually be in imperative form ("Do") and short.
For example:
`// Do the calculation`
`// Calculate`
They can also be short, so you could write:
`// Do calculation`
Instead of:
`// Do the calculation`

In this case, I would write it as:
`// Align x coordinate to be symmetric with respect to y-axis`

File: public/view/view.js [hunk 13]
```diff
@@ -111,44 +120,115 @@ function navigateGrid(event) {
...
+    // as explained before for x coordinate
+    if (rowCount % 2 == 0) {
+        yCoordinate += 0.5;
+    }
+    // We negate the value of y because we want uper cells to have positive values of y
```
Comment (oferb) [inline public/view/view.js]: I would shorten:
`// We negate the value of y because we want uper cells to have positive values of y and lower cells with negative values`
To:
`// Negate y so upper cells to have positive values and lower cells have negative values`
This could even be shortened to (assuming it's clear what happens to the other side):
`// Negate y so upper cells to have positive values`

File: public/view/view.js [hunk 14]
```diff
@@ -111,44 +120,115 @@ function navigateGrid(event) {
...
 }
 
-function startSoundPlayback(event) {
-    currentCellUnderTouchPoint = event.currentTarget;
-    event.preventDefault();
-    stopSoundPlayback(event);
-    let selectedValue = event.currentTarget.firstChild.data;
-    playSound(selectedValue);
+/**
+* This function calculates the greatest Euclidean distance possible for a cell in the grid,
```
Comment (oferb) [inline public/view/view.js]: No need to write "This function".
Take a look at the documentation here:
https://docs.oracle.com/javase/8/docs/api/java/util/List.html

They all start with words like:
Returns
Removes
Retains
Replaces

These are all present simple third person:
https://www.worddive.com/grammar/en/english-grammar/5-present-simple-third-person/

So here it would be:
`Calculates the greatest Euclidean distance possible for a cell in the grid, ...`
Also later,
`Assumes that the number of cells...`

File: public/view/view.js (hunk 13 above)
Comment (mohammad-suliman) [reply public/view/view.js]: I prefer the longer version suggested. What do you think?

Comment (oferb) [reply public/view/view.js]: Sure

File: public/view/view.js [hunk 15]
```diff
@@ -111,44 +120,113 @@ function navigateGrid(event) {
...
-function getChildIndex(currentChild) {
-    let index = 0;
-    while (currentChild.previousSibling) {
-        index++;
-        currentChild = currentChild.previousSibling;
+/**
+* Maps the row and col attributes for each cell in the grid 
+* to 2D coordinates (x,y)
+* Row index is mapped to y coordinate, and col index is mapped to x coordinate
+* The row and col attributes are zero based indexes
```
Comment (oferb) [inline public/view/view.js]: I think you can shorten this to:
```
/**
Maps each cell's row and col to a 2D coordinate. Returns a map with the x and y coordinates (e.g {x:1, y:0}).
Examples:
Cells in a 3X3 grid will be positioned between -1 and 1 in both x and y.
Cells in a 2X2 grid will be positioned between -0.5 and 0.5 in both x and y.
*/
```

This is an implementation detail so should be inside the method:
`* Note: row and col are zero based indexes`

File: public/view/view.js [hunk 16]
```diff
@@ -111,44 +120,113 @@ function navigateGrid(event) {
...
+    playSound(event);
+}
+
+function playSound(event) {
+    if (audioContext.state == "suspended") {
+        audioContext.resume();
+    }
+    let currentCell = selectedCell;
+    // Create oscillator and panner nodes and connect them each time we want to play audio
+    // because those nodes are singel use intities
```
Comment (oferb) [inline public/view/view.js]: `intities->entities`
